### DISCUSSION TITLE
HttpRunner 4.0 实现 golang/python 双引擎

### DISCUSSION THREAD

**Discussion by debugtalk** [httprunner/httprunner#1186] [1 - 想法（Ideas）]
> 当前还只是非常初步的想法，有极大的可能会进行调整

当前 HttpRunner+ 采用 golang 重新进行了设计和开发，因此重新开了一个新的仓库 `httprunner/hrp`，跟原有的 python 版 `httprunner/httprunner` 有些割裂。

也许，我们可以将 Python 版的 HttpRunner 和 Golang 版的 HttpRunner+ 进行融合，实现 HttpRunner 4.0？

初步想法包括：

- 代码仓库合并为一个，都在 `httprunner/httprunner` 中维护
- 将命令行工具统一为 `hrp`，采用 golang 开发
- 生成脚本语言的功能统一在 `hrp har2case` 中实现，支持生成 YAML/JSON/Golang/Python 格式的脚本
- 新增 `hrp init --py` 命令，可初始化 Python3 venv，并在 venv 中安装 Python 版的 HttpRunner
- 在执行测试的时候，`hrp run` 作为统一执行的入口：
  - 指定脚本路径为 YAML/JSON 格式时，采用 `hrp` golang 内置实现的执行引擎
  - 指定脚本路径为 Golang 格式时，在 golang 代码中调用 `go test` 执行测试
  - 指定脚本路径为 Python 格式时，在 golang 代码中调用 `pytest` 执行测试

**Comment by huangbin-scu** [comment]
不同格式脚本的特性一致性问题：因为python和go的脚本的底层执行引擎不一样，那python和go版的脚本里支持的特性的一致性维护可能需要注意下~从外层看合到了一起，但是新增一个新特性是不是在python和go里都得实现一次，维护成本会比较高吧

**Reply by debugtalk** [reply]
1、关于重复实现的问题：绝大多数功能都会采用 golang 实现，例如命令行工具（`hrp`）、脚本生成（`har2case`）等；在执行引擎层，如果是执行 YAML/JSON 的用例，也是采用 golang 实现的；只有针对 Golang/Python 脚本类型，执行引擎是分开实现的，如果存在变更或功能新增，的确是需要同时调整。
2、一致性保障：会在 CI 里面加上用例，确保迭代过程中的兼容性。

**Comment by imath60** [comment]
隆哥，可以看下 gatling 和 k6 的玩法，现在主流的压测引擎是底层用高性能的并发模型（actor 或 csp），上层多语言支持 java系（java/kotlin/scala/js）或 python系，通过插件扩展的方式支持多协议 feeder assert 自定义处理函数等

好的性能测试引擎是一定可以做功能测试的【通过扩展插件 丰富断言类型 丰富测试数据源这些】

底层并发模型的选择问题，如果选 actor 模型，看 gatling；如果是 csp 模型，开源实现看 k6，现在已经被 Grafana 收购了。

**Reply by debugtalk** [reply]
@imath60 当前 HttpRunner 也基本上是按照这个思路在进行实践。你推荐的两个模型我再看下。

**Comment by ZhouYixun** [comment]
可以往上层实现

**Comment by eeslook** [comment]
“采用 golang 重新进行了设计和开发”的初衷是什么？为什么不是往其他类型的runner方向、性能测试、提升自动化效率等方向扩展，而选择切换成go语言？

**Reply by debugtalk** [reply]
@eeslook 这里可能就是 Golang 静态语言 和 Python 动态语言的选择了；选择 Golang 的原因主要有 2 个：

1、Golang 是静态、强类型语言，可维护性更高；
2、我们期望在 hrp 中将压测能力的单机发压能力做得更高，数据更准确；基于 Python 的 locust 是比较难达成这个目标。

而你说的「其他类型的runner方向、性能测试、提升自动化效率等方向」，跟切换成 golang 并不冲突，而且切换成 golang 以后可以帮助 HttpRunner 更好地达成这些目标。

**Reply by eeslook** [reply]
测试行业普遍在python语言、java语言，go语言目前使用仍然是比较少，在采用的技术上，多数公司都会考虑维护成本，使用go会增加这方面风险，在维护性上会推高门槛，个人愚见。
极致的简单好用（包含api和UI的被集成能力），作为测试行业的独角兽类型的解决方案，能够解决核心的测试诉求会是比较好的方向？

**Reply by debugtalk** [reply]
对于 HttpRunner，目标是跟具体的编程语言解耦，在测试用例脚本层面会支持多种主流的编程语言，Go/Python/Java 都会支持。对于 hrp 的用户来说，应该是感受不到 hrp 是采用什么语言开发的，他可以选择自己喜欢的脚本格式和语言。

**Reply by csh589** [reply]
但很多已经成型了的用例库是使用pytest的方式编写。。比如我们这里，在pytest形式写用例的基础上，重度依赖了pytest的很多特性。一旦要做底层框架和语言的迁移，成本是非常非常高的。
其实不太理解为什么用两种语言然后放在了同一个库里面。。看文档还会保持两边的特性对齐？这个工作量不止double的样子

**Comment by xyd0000** [comment]
不知发展方向是重接口自动化测试还是压力性能测试，还是能同时兼顾两者。本人不是从事互联网，压力测试做的不多也不大。感觉可以考虑一下项目使用的用户。

**Reply by debugtalk** [reply]
目标是两者兼顾的。可以看下设计理念的介绍：
https://httprunner.com/docs/overview/

**Comment by ZhouYixun** [comment]
参考appium的形式，用户不会在意编写用例是什么语言或下层逻辑。
假如httprunner4.0出来了，那么用户可以用py或go编写用例，直接给httprunner4.0去执行。
httprunner4.0会根据用例自动调用httprunner或hrp，或者会本地起个http服务给用户塞用例进去也是可以的。
我的理解是直接往上做就行

**Reply by debugtalk** [reply]
appium 的形式是实现一个 server，再用不同的语言实现 client，用户编写脚本的时候是选择某一种语言跟对应的 client 进行交互。

hrp 这边有考虑过这种思路，不过当前采用的方式还是基于 hashicorp/plugin 的方式。针对 YAML/JSON 这类脚本类型是最好做的，方便采用 plugin 实现各种语言类型的插件。

而针对 Golang/Python 这些脚本类型，因为涉及到语言运行时，因此需要调用特定语言的运行时，初步想法是在外层包一层调用。

**Comment by xulz** [comment]
之前没用过HttpRunner，最近体验了下v3和v4版本，个人的一些看法：
1.  v3使用Python开发，录制或postman导出的json/yaml格式会在执行时被转换为pytest格式，文档也强烈推荐使用pytest格式，是出于最终一致性和方便扩展考虑吧。
2.  V4使用Go开发，json/yaml格式会在执行时被转换为Go对象后直接执行。
3.  如果能支持Python双引擎，感觉会更好，不至于同时维护两个版本。

从使用者的角度，python的用户基础更大，对于只有python经验的人初次上手go开发环境并不足够友好（使用国内镜像等）。

从定位的角度，Go的性能优势很明显，之前我用locust时每次都要开好几个slave。另外Go的二进制文件分发和测试环境部署也会更方便。

能很好的融合是个不错的思路。


补充：
试了下执行v3的示例yaml，排除自定义函数部分的错误后，validate部分v4也不太兼容，v4现在是：
```
check: status_code
assert: eq
expect: 200
```
v3是`eq: ["status_code", 200]`.

从代码实现看，v4的格式与v3的format1类似。

**Reply by debugtalk** [reply]
@xulz 格式兼容这块儿，在最新的 hrp 版本中已经实现兼容了。
